### ISSUE TITLE
New omnipay processor - recurring notes?

### ISSUE THREAD

**Issue by adixon** [eileenmcnaughton/nz.co.fuzion.omnipaymultiprocessor#19]
I've got a new payment processor here using someone's recently developed omnipay plugin, for BluePay
https://github.com/zburke/omnipay-bluepay

Seems to be working so far (after I finished reading all the instructions), but I can't see many hints about making the recurring function working. I seem to recall that you implemented it using tokens, and I can see that we need at least a createCard method (because I get an error that it's missing).

What's your best example of a createCard function?

**Comment by eileenmcnaughton** [comment]
The one I worked on just extends the main function - ie. it just passes out slightly different parameters.

https://github.com/eileenmcnaughton/nz.co.fuzion.omnipaymultiprocessor/blob/master/vendor/omnipay/paymentexpress/src/Message/PxPayCreateCardRequest.php

**Comment by adixon** [comment]
Okay, I've made some progress - I can fill out the contribution form with a recurring value and it gets processed over on BluePay. But it gets left as pending in CiviCRM and no token is getting created.

It appears that the processPaymentResponse function should be responsible for completing the transaction and creating the token, but I can't see where that is getting called (or should be getting called) and whether there's something I'm missing or not returning properly.

Any more clues?

**Comment by eileenmcnaughton** [comment]
civicrm/payment/ipn/xx is the notify url. That results in a call with whatever is in the parameters to handleNotification and I think the CompletePaymentResponse gets called

https://github.com/eileenmcnaughton/nz.co.fuzion.omnipaymultiprocessor/blob/master/vendor/fuzion/omnipay-paybox/src/Message/SystemCompletePurchaseRequest.php

(note System is probably a prefix specific to that gateway)

**Comment by adixon** [comment]
Hmm, cryptic ... so it seems you're saying that the only way that recurring payments are completed is via an (external call?) to that notify url, and maybe that means that recurring contributions with tokens have to use billing_mode = 4? That doesn't sound right.  How else is that notify url going to get called? Is there an omnipay message that I need to implement? Or does it depend on some standard notification service that omnipay implements?

**Comment by eileenmcnaughton** [comment]
Sorry - first question is this a recurring token?

**Comment by eileenmcnaughton** [comment]
ie driven by the cron job

**Comment by eileenmcnaughton** [comment]
The cron job (which obviously needs enabling) creates a pending recurring contribution. Contacts the processor to process it and then completes it

https://github.com/eileenmcnaughton/nz.co.fuzion.omnipaymultiprocessor/blob/master/api/v3/Job/ProcessRecurring.php#L37

The api paymentProcessor.pay simply calls doPayment

https://github.com/eileenmcnaughton/nz.co.fuzion.omnipaymultiprocessor/blob/master/api/v3/PaymentProcessor/Pay.php#L14

Is it failing in the pay api?

**Comment by adixon** [comment]
Ah, okay - so the flow is different that I understood, i.e. the completion of the first payment doesn't happen right away, not until the first recurring job.

In other words, it's all working as designed ... thanks.

**Comment by adixon** [comment]
Hmm, nope, that didn't work.

Specifically, running the cron job doesn't do anything. Re-reading your comment, maybe I'm not explaining it. The issue is that the initial contribution and the recurring contribution series are both left in pending status (which is the civi default, right?). Running the cron job doesn't do anything. I did notice when editing the recurring contribution that no next scheduled date is set, presumably due to the pending status, and that no token was created.

So I guess the mystery is still: when a recurring contribution (series + initial contribution) is created from  a public donation form, how is the series supposed to get approved and the token created? i.e. what triggers the processPaymentResponse?

**Comment by eileenmcnaughton** [comment]
OK so the createCard function needs to 

1) create the recurring entity - which in the example I pasted to is basically the same as doing a purchase but passing one more parameter. For other processors the difference is greater.

2)  process the payment provided the action is Purchase - which it is when we call it.

3) return an appropriate Request object.

e.g from eway

    /**
     * Store a credit card as a Token
     *
     * You can currently securely store card details with eWAY for future
     * charging using eWAY's Tokens.
     * After storing the card, pass the cardReference instead of the card
     * details to complete a payment.
     *
     * @link https://eway.io/api-v3/#create-token-customer
     * @param array $parameters
     * @return \Omnipay\Eway\Message\RapidDirectCreateCardRequest
     */
    public function createCard(array $parameters = array())
    {
        return $this->createRequest('\Omnipay\Eway\Message\RapidDirectCreateCardRequest', $parameters);
    }

The request object's sendData function returns a Response object

    public function sendData($data)
    {
        $httpResponse = $this->httpClient->post($this->getEndpoint(), null, json_encode($data))
            ->setAuth($this->getApiKey(), $this->getPassword())
            ->send();

        return $this->response = new RapidResponse($this, $httpResponse->json());
    }

The response object needs to implement isSuccessful - e.g

    public function isSuccessful()
    {
        return isset($this->data['TransactionStatus']) && $this->data['TransactionStatus'];
    }


Here is how our extension calls it
    try {
      if (!empty($params['is_recur'])) {
        $response = $this->gateway->createCard($this->getCreditCardOptions(array_merge($params, array('action' => 'Purchase')), $component))->send();
      }
      else {
        $response = $this->gateway->purchase($this->getCreditCardOptions($params, $component))
          ->send();
      }
      if ($response->isSuccessful()) {
        // mark order as complete
        $params['trxn_id'] = $response->getTransactionReference();
        //gross_amount ? fee_amount?
        return $params;
      }
      elseif ($response->isRedirect()) {

      }

**Comment by adixon** [comment]
Yup, and all that is working but there's still no token being stored in CiviCRM and the records are still in pending. I can see the trxn_id being returned successfully in the params, and I can see that the token is being created over on the payment processor.

And, I still can't see where in that flow the processPaymentResponse might be getting called, assuming that's how the token should be getting created, which might not be a correct assumption.

As I look at your code, I wonder ... it looks to me like you're giving up after creating the credit card but before generating a charge with it?

Specifically, I'd expect a second message to go out after the card/token is created, requesting a payment with the card. Unless by 'extra parameter' you are modifying the Omnipay concept of createCard and allowing it to run a charge at the same time? No, that would be confusing and I can't see where you'd have done this anyway in the vendor branches ...

Okay, I'm going to test my theory and add an extra line to actually run a payment with the new card in there.

**Comment by eileenmcnaughton** [comment]
"Unless by 'extra parameter' you are modifying the Omnipay concept of createCard"

Right - the createCard does process a payment for PaymentExpress if the 'action' is purchase. I think that's a supported concept with CreateCard

**Comment by adixon** [comment]
Okay, I've taken this up again, and might be closer to understanding it. I've got a pending recurring contribution being created and the individual contribution completed and the contribution on the payment processor looking good. I'm not getting the recurring record completed, nor the token being stored. So I'm guessing that has to happen via an IPN call.

But ... since I'm on call waiting to talk to the payment processor about adding an ipn callback to the account, I'm also wondering why we can't do it with the result we get back from the initial request, assuming it's validated. I'm guessing that you implemented this for direct post, where it has to be an IPN to validate the payment, but since my estimated wait time is increasing instead of decreasing, I'm going to try patching the payment object to allow this.

**Comment by JoeMurray** [comment]
@adixon what was the resolution of this issue? Just wondering as I consider implementing recurring for a different payment processor using OmniPay.

**Comment by adixon** [comment]
I implemented this with some additions to the extension, here: https://github.com/adixon/nz.co.fuzion.omnipaymultiprocessor
I wasn't super happy with how I did it and don't recommend it.

Working with omnipay is hard - it has a different code style and an extra level of abstraction that made understanding and debugging code difficult (for me!).
